blog/2020/09/rr-memory-magic.md

```diff
@@ -0,0 +1,503 @@
+
+Alright, so looking at the event log, we basically see reflected what we already knew: It does a bunch of write system calls (to print the episode number), a ton of computation (143 billion conditional branches' worth), and eventually it's supposed to crash. We also know that at event `51909` (the last write before the supposed segfault), things in the replay were still mostly on track. We can't say whether the memory contents had already diverged, but we at least know that the register state and the number of instructions executed was bitwise identical to what happened during the recording.
+
+One good way to investigate a divergence is to poke around in memory a bit prior to when one suspects the divergence to
+have occurred and see what kind of memory modifications could have made the difference. Usually, I just do this by going to the prior event, but in our case, there are 15 million conditional branches between the last two events, which is quite a bit of distance. Instead I hacked up a [quick command](https://github.com/mozilla/rr/pull/2658) that lets us seek directly to a particular tick count. In this case, I arbitrarily chose `143998750000`, about 10000 ticks prior to when the segfault occurred during the recording. Let's go there:
+
+```
+$ rr replay -g 51908 --serve-files .
+[snip]
+0x0000000070000002 in ?? ()
+(rr) when
+Current event: 51909
+(rr) when-ticks
+Current tick: 143983372944
+(rr) seek-ticks 143998750000
+
+(rr) when-ticks
+Current tick: 143998750000
+(rr) when
+Current event: 51911
+
+(rr)
+```
+
+The trace tells us that it expected the segfault to happen at address `rip:0x7f049e2ba8e3`, so let's take a look there:
+
+```
+(rr) disas 0x7f049e2ba8e3
+Dump of assembler code for function jl_apply_tuple_type:
+[snip]
+   0x00007f049e2ba8e0 <+96>:  mov    (%rax),%r10
+   0x00007f049e2ba8e3 <+99>:  mov    -0x8(%r10),%r9
+[snip]
+```
+
+Alright, so that's a load from `r10`. Luckily `r10` is recorded in the trace: `r10:0x1007f0490b784d0`. A seasoned debugger will quickly recognize this pointer as too wide. In particular, this pointer has the lowest bit of its high byte set. These bits are mostly unused. Sometimes unused pointer bits are used for extra information, but at least julia itself only uses the unused low bits for this purpose, not the high bits. That said, this trace included a significant number of external native libraries (including a full python environment), so it's certainly possible that some external library would have used such pointer tagging techniques. Let's do some more investigation. The first thing I tried to do was take a look at the pointer with the high bit manually cleared to see if it was a valid julia object at all or just some junk that somehow ended up being loaded from:
```

Comment:
```suggestion
Alright, so that's a load from `r10`. Luckily `r10` is recorded in the trace: `r10:0x1007f0490b784d0`. A seasoned debugger will quickly recognize this pointer as too wide. In particular, this pointer has the lowest bit of its high byte set. These bits are mostly unused. Sometimes unused pointer bits are used for extra information, but at least Julia itself only uses the unused low bits for this purpose, not the high bits. That said, this trace included a significant number of external native libraries (including a full Python environment), so it's certainly possible that some external library would have used such pointer tagging techniques. Let's do some more investigation. The first thing I tried to do was take a look at the pointer with the high bit manually cleared to see if it was a valid julia object at all or just some junk that somehow ended up being loaded from:
```